lib/linux_kernel/kthread.c

```diff
@@ -207,9 +208,16 @@ M0_INTERNAL int m0_thread_confine(struct m0_thread *q,
 	if (!zalloc_cpumask_var(&cpuset, GFP_KERNEL))
 		return M0_ERR(-ENOMEM);
 
-	for (idx = 0; idx < nr_bits; ++idx)
-		if (m0_bitmap_get(processors, idx))
+	for (idx = 0; idx < nr_bits; ++idx) {
+		if (m0_bitmap_get(processors, idx)) {
 			cpumask_set_cpu(idx, cpuset);
+			if (idx1 == -1)
+				idx1 = idx;
+		}
+	}
+	M0_ASSERT(idx1 != -1); /* At least one processor must be set. */
+	q->t_tls.tls_loc = idx1;
```

Comment:
[* question *] Do I understand correctly that an assumption is used here about how localities are assigned to CPU cores? If so, then it's a wrong place to make this assignment. It should be done near the place where the decision about which CPUs to run locality thread on is taken.

Reply:
`m0_thread_confine()` is called as a result of such decision. We just remember this result at TLS here.

Reply:
`m0_thread_confine()` could be called from a place other than locality thread confining. The point is that the locality-specific code is added to a function that has nothing to do with localities.

Reply:
Ah, this might be a confusion caused by the `tls_loc` name, which stores the cpu id, actually, not the m0_locality. Le me rename it to `tls_cpu_id`.

Reply:
This wouldn't explain how the CPU id is chosen assuming there could be many CPUs in the bitmap.

Reply:
As @nikitadanilov [suggested](https://github.com/Seagate/cortx-motr/pull/833#discussion_r665408169), we "reset it to the first processor in the bitmap". This cpu id is used only to select the locality from the array of localities in the domain at `m0_locality_here()` (via `m0_locality_get()`).

Reply:
> `m0_thread_confine()` could be called from a place other than locality thread confining.

@max-seagate, I'm not sure if it is possible. This might lead to the same panic mentioned at the top.